Azure/azure-cli
# [Core][Profile] Add `tenant` parameter to `get_raw_token`

Inspired by #11351, #11780

**In cli core**: `get_raw_token`: Add `tenant` parameter to acquire token for the tenant directly, needless to specify a subscription. The structure of the return value is not changed - a `(cred, subscription, tenant)` tuple, but the second element of the tuple - `subscription` is set to `None` for tenant token for backward compatibility. 

**In command module**: `az account get-access-token`: Add `--tenant` parameter to acquire token for the tenant directly, needless to specify a subscription. In that case, the returned JSON doesn't have `subscription` field.

```jsonc
{
  "accessToken": "eyJ0e...",
  "expiresOn": "2020-01-08 17:09:52.554667",
  "tenant": "72f988bf-86f1-41af-91ab-2d7cd011db47",
  "tokenType": "Bearer"  // no subscription if --tenant is specified**
}
```

The returned token can be decoded at https://jwt.ms/ for verification. This is a Microsoft website so there is no security concern to paste the token there. Check the `tid` field for the tenant ID. 

`tenant` is only available for user and service principal account, not for MSI or Cloud Shell account, because MSI or Cloud Shell account is a single-tenant account and it is meaningless to explicitly specify a tenant. We error it out for simplicity, otherwise we need to check whether the specified tenant is the home tenant of the account.

To test:

* Test help: `az account get-access-token -h`
* Test with user identity: 
    ```sh
    az login
    account get-access-token --tenant {}
    ```
* Test with service principal:
    ```sh
    az ad sp create-for-rbac
    az login --service-principal -u {} -p {} -t {}
    az account get-access-token --tenant {}
    ```
* Test with MSI (this will fail!): create a VM with MSI and setup the dev environment, then run 
    ```sh
    az login --identity
    az account get-access-token --tenant {}
    ```
* Cloud Shell can't be tested currently, so unit test `test_get_raw_token_in_cloud_console` is added to mock Cloud Shell

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).
